snapcraft/project/_project.py

```diff
@@ -84,9 +84,25 @@ def _get_build_base(self) -> str:
         return self._snap_meta.get_build_base()
 
     def _get_project_directory_hash(self) -> str:
-        m = hashlib.sha1()
-        m.update(self._project_dir.encode())
-        return m.hexdigest()[:6]
+        # This function uses md5 hashes because they are fast, and because
+        # the hashes only need to be unique per project, so clashes are
+        # tremendously unlikely and not a big deal even if they happen.
+        hashes: List[str] = list()
+        for root, dirs, files in os.walk(self._project_dir):
+            # Sort contents, we need this to be stable so it's reproducible
+            dirs.sort()
+            files.sort()
+
+            for filename in files:
```

Comment:
do you have impact numbers for this? What's the hit on a project like Snapcraft itself which is not so small neither so large?

Reply:
Sure, here's a test script:

```python3
#!/usr/bin/env python3

import time

import snapcraft.project

if __name__ == "__main__":
    project = snapcraft.project.Project()
    start = time.time()
    h = project._get_project_directory_hash()
    end = time.time()

    print(f"{h}: calculated in {end - start:.2f} seconds")
```

Running it in the snapcraft dir:

```
$ ./test.py 
f41210f4761dc6bafbb01b1cb9ab7416: calculated in 0.12 seconds
```

Here's the Nextcloud snap:

```
$ ./test.py 
eb634d5edcc2a5deaada38e330197f2f: calculated in 0.01 seconds
```